### PR TITLE
feat: add support for dataset built-in labels via the label key in HIF processing

### PR DESCRIPTION
**Pull request by ddevin96**
### All submissions

- [x] Have you followed the guidelines in our [Contributing](../../CONTRIBUTING.md) document?
- [x] Have you checked to ensure there are no other open [Pull Requests](../../../pulls) for the same changes?
- [x] Have you made sure you have rebased your branch to the latest commit on the target branch?

### Description

- Added support for datasets' built-in labels via the `label` key in HIF processing.
- Added method to `Dataset` to convert label tensor to human-readable strings.

### Checklist

- [x] Does your submission pass all tests? (use `make test`)
- [x] Have you written tests to cover all your changes? If not, provide a reason.
- [x] Have you lint your code locally before submission? (use `make format`)
- [x] Have you type checked your code locally before submission? (use `make typecheck`)

### CODE REVIEW

File: hypertorch/data/dataset.py [hunk 1]
```diff
@@ -80,6 +80,9 @@ def __init__(
         validate_task(self.task)
 
         self.hdata: HData = hdata if hdata is not None else HData.empty(task=task)
+        self.hif_hypergraph: HIFHypergraph | None = (
```
Comment (tizianocitro) [inline hypertorch/data/dataset.py]: We should remove this and keep only `self.__hif_hypergrap` at line 74.

File: hypertorch/data/dataset.py [hunk 2]
```diff
@@ -664,6 +667,14 @@ def to(self, device: torch.device) -> Dataset:
         self.hdata = self.hdata.to(device)
         return self
 
+    def to_human_readable_y(self, y: Tensor) -> list[str]:
+        """
+        Convert numeric labels in `y` to human-readable string labels using the label map
```
Comment (tizianocitro) [inline hypertorch/data/dataset.py]: Missing Args and Returns. Probably also ass a Raises tag for the `self.hif_hypergraph` that can raise when None.

File: hypertorch/data/hif.py [hunk 3]
```diff
@@ -159,10 +170,39 @@ def process_hypergraph(
...
+            y=y,
         )
 
         return (hdata, hif_hypergraph)
 
+    @classmethod
+    def __process_y(
+        cls,
+        hypergraph: HIFHypergraph,
+        num_nodes: int,
```
Comment (tizianocitro) [inline hypertorch/data/hif.py]: I believe this should be called `size` instead of `num_nodes`.

File: hypertorch/data/hif.py [hunk 4]
```diff
@@ -159,10 +170,39 @@ def process_hypergraph(
...
+        dtype: torch.dtype,
+    ) -> tuple[Tensor | None, dict[str, int | float] | None]:
+        """
+        Build the node label tensor from HIF node attributes.
```
Comment (tizianocitro) [inline hypertorch/data/hif.py:186]: We should complete the docstring.

File: hypertorch/data/hif.py [hunk 5]
```diff
@@ -159,10 +170,39 @@ def process_hypergraph(
...
+            and len(list_attr_label) == num_nodes
+        ):
+            sorted_unique_labels = sorted(set(list_attr_label))
+            map_label_to_index: dict[str, int | float] = {
+                label: idx for idx, label in enumerate(sorted_unique_labels)
+            }
+            y = torch.tensor([map_label_to_index[label] for label in list_attr_label], dtype=dtype)
+            return y, map_label_to_index
+
+        raise ValueError("Node label attribute is missing for some nodes in the hypergraph.")
```
Comment (tizianocitro) [inline hypertorch/data/hif.py:217]: Not now but we can create an issue for this raise to include the exact indexes that have no label.

File: hypertorch/tests/data/dataset_test.py [hunk 6]
```diff
@@ -2017,3 +2047,68 @@ class TestDataset(Dataset):
...
+):
+    with patch.object(
+        HIFLoader,
+        "load_by_name",
+        return_value=(mock_hdata_with_labels, mock_hypergraph_with_mixed_labels),
+    ):
+        dataset = AlgebraDataset(task=task)
+
+    readable = dataset.to_human_readable_y(dataset.hdata.y)
+    print(readable)
```
Comment (tizianocitro) [inline hypertorch/tests/data/dataset_test.py]: Remove the print. This should assert type and len of `readable`, I suppose.

File: hypertorch/tests/data/dataset_test.py [hunk 7]
```diff
@@ -2017,3 +2047,68 @@ class TestDataset(Dataset):
         assert torch.allclose(
             result, torch.tensor([1.5, 10.0], dtype=torch.float)
         )  # capacity, weight (insertion order)
+
+
+def test_property_hif_hypergraph_returns_correct_values(mock_hdata, mock_hif_hypergraph):
```
Comment (tizianocitro) [inline hypertorch/tests/data/dataset_test.py]: I believe we already have a test like this at line 207. Should we merge them or keep only one?

File: hypertorch/tests/data/dataset_test.py [hunk 8]
```diff
@@ -2017,3 +2047,68 @@ class TestDataset(Dataset):
...
+
+    hif = dataset.hif_hypergraph
+    assert hif.metadata == mock_hif_hypergraph.metadata
+    assert hif.nodes == mock_hif_hypergraph.nodes
+    assert hif.hyperedges == mock_hif_hypergraph.hyperedges
+    assert hif.incidences == mock_hif_hypergraph.incidences
+    assert hif.network_type == mock_hif_hypergraph.network_type
+
+
+def test_property_hif_hypergraph_raises_error_when_hif_hypergraph_is_not_loaded(
```
Comment (tizianocitro) [inline hypertorch/tests/data/dataset_test.py]: Same, we already have this at line 220. This should be removed.

File: hypertorch/tests/data/dataset_test.py (hunk 7 above)
Comment (tizianocitro) [reply hypertorch/tests/data/dataset_test.py]: For ordering reason, we should move these tests above in case.

File: hypertorch/tests/data/dataset_test.py [hunk 9]
```diff
@@ -2017,3 +2047,68 @@ class TestDataset(Dataset):
...
+    mock_hdata, mock_hif_hypergraph
+):
+    with patch.object(HIFLoader, "load_by_name", return_value=(mock_hdata, None)):
+        dataset = AlgebraDataset()
+
+    with pytest.raises(ValueError, match=re.escape("HIF hypergraph is not available.")):
+        _ = dataset.hif_hypergraph
+
+
+def test_set_hif_hypergraph_sets_hif_hypergraph(mock_hdata, mock_hif_hypergraph):
```
Comment (tizianocitro) [inline hypertorch/tests/data/dataset_test.py]: We can keep this but should be moved above.

File: hypertorch/tests/data/dataset_test.py (hunk 7 above)
Comment (tizianocitro) [reply hypertorch/tests/data/dataset_test.py]: I suppose the ones we have now have bad naming. For example: `test_dataset_property_returns_correct_values` and `test_dataset_property_raise_error_when_hif_hypergraph_is_none`.

We should rename those anyway.

File: hypertorch/tests/data/dataset_test.py (hunk 8 above)
Comment (tizianocitro) [reply hypertorch/tests/data/dataset_test.py]: See this [message](https://github.com/hypernetwork-research-group/hypertorch/pull/360/changes#r3577085135).

File: hypertorch/tests/data/hif_test.py [hunk 10]
```diff
@@ -1031,3 +1051,36 @@ def test_hifloader_download_failure_when_hf_token_is_invalid(tmp_path):
...
+    hdata, hif_hypergraph = HIFProcessor.process_hypergraph(
+        mock_hypergraph_with_labels, task=TaskEnum.NODE_CLASSIFICATION
+    )
+    assert hif_hypergraph is not None
+    assert hif_hypergraph.metadata == {
+        "label_map": {"A": 0, "B": 1},
+    }
+    assert hdata.y is not None
+    assert hdata.y.shape[0] == hdata.num_nodes
+    assert torch.all(hdata.y == torch.tensor([0, 1], dtype=torch.long))
```
Comment (tizianocitro) [inline hypertorch/tests/data/hif_test.py]: Is it more clear if we do:

```python
expected = torch.tensor([0, 1], dtype=torch.long)
assert torch.equal(hdata.y, expected)
```

I'm not sure why we need `all()` as `[0,1]` should be exactly the tensor `y`.

File: hypertorch/tests/data/hif_test.py [hunk 11]
```diff
@@ -1031,3 +1051,36 @@ def test_hifloader_download_failure_when_hf_token_is_invalid(tmp_path):
...
+            mock_hypergraph_with_missing_labels, task=TaskEnum.NODE_CLASSIFICATION
+        )
+
+
+def test_process_hypergraph_without_labels(mock_sample_hypergraph):
+    hdata, _ = HIFProcessor.process_hypergraph(
+        mock_sample_hypergraph, task=TaskEnum.NODE_CLASSIFICATION
+    )
+
+    assert torch.all(hdata.y == torch.tensor([1, 1], dtype=torch.long))
```
Comment (tizianocitro) [inline hypertorch/tests/data/hif_test.py]: Same as [above message](https://github.com/hypernetwork-research-group/hypertorch/pull/360/changes#r3577106890).

**Review by tizianocitro — CHANGES_REQUESTED** [review]

File: hypertorch/tests/data/dataset_test.py [hunk 12]
```diff
@@ -2017,3 +2047,39 @@ class TestDataset(Dataset):
...
+):
+    with patch.object(
+        HIFLoader,
+        "load_by_name",
+        return_value=(mock_hdata_with_labels, mock_hypergraph_with_mixed_labels),
+    ):
+        dataset = AlgebraDataset(task=task)
+
+    readable = dataset.to_human_readable_y(dataset.hdata.y)
+    assert type(readable) is list
```
Comment (tizianocitro) [inline hypertorch/tests/data/dataset_test.py:2084]: This should assert the type of the elements in `readable`. For example, for `TaskEnum.NODE_CLASSIFICATION` should assert `int`.

Comment (tizianocitro) [reply hypertorch/tests/data/dataset_test.py:2084]: My bad, this is wrong.

**Review by tizianocitro — APPROVED** [review]
